Add weekly Slack activity summary workflow (#1366)
C2SM/icon4py: scripts/data/weekly_slack_summary_instructions.md

# Weekly Slack Summary Instructions

You are generating a concise weekly activity summary for the icon4py repository.
The summary will be posted to a Slack channel as `mrkdwn` text.

Use the provided Markdown context file (`weekly_slack_summary_context.md`) as the
source of facts.

## Ground rules

- Write the final summary to the file path given in the prompt. This file is

>   what gets posted to Slack - do not output the summary to stdout.

Author: ```suggestion
- Write the final summary to the file path given in the prompt. This file is
  what gets posted to Slack.
```

- Use only facts present in the context.

> - Do not use em-dashes (`—`); use hyphens (`-`) or colons instead.

Author: ```suggestion
- Do not use em-dashes (`—`); use simple sentences separated by periods (`.`) or if necessary joined by colons (`:`) or semicolons (`;`).
```

- Do not use Markdown headings (`#`, `##`, etc.). Use `*Section Title*` for
  section titles (Slack renders single asterisks as bold).
- Format every link as Slack mrkdwn: `<URL|display text>`. Do not use raw URLs
  or Markdown `[text](url)`.
- Use `-` as list marker.

## Length limit

**The summary must be under 3,800 characters and under 50 lines.**
Slack's hard limit is 4,000 characters. Stay under 3,800 to leave a safety

> margin. Be informative - use the full budget, but do not exceed 3,800.

Author: ```suggestion
margin. Be informative: use the full budget, but do not exceed 3,800.
```

After writing the summary to the file, verify its length with a shell command
(e.g. `wc -m <path>`). If it exceeds 3,800 characters, edit the file to trim
content (drop descriptions before titles, reduce entry counts), then verify

> again. Repeat until it fits. Never guess the length - always measure it.

Author: ```suggestion
again. Repeat until it fits. Never guess the length, always measure it.
```

Guidelines:

- Merged PRs: list all merged PRs (combine rest into a group bullet if >10)

> - Closed PRs: one group bullet (no individual entries, no URLs)

Author: ```suggestion
- Closed PRs: list all PRs closed without merging (combine rest into group bullet if >5)
```

> - Ongoing PRs: list notable ongoing PRs (up to ~10)

Author: ```suggestion
- Ongoing PRs: list notable ongoing PRs (up to ~10), ongoing means commits in the last two weeks
```

> - Inactive PRs: up to 5 entries (no descriptions)

Author: ```suggestion
- Inactive PRs: list notable inactive PRs (up to ~5), inactive means no commits in the last two weeks
```

- Issues: list all, keep short
- CI sections: one line each, include pipeline URL

## Output format

1. `*icon4py weekly summary: <start-date> - <end-date>*`
2. `*Merged PRs*`
3. `*Closed PRs (without merging)*`: Omit section if none.
4. `*Ongoing PRs*`
5. `*Inactive open PRs*`: Mention total count of unlisted ones.
6. `*Opened issues*`
7. `*Closed issues*`
8. `*Weekly CI status*`
9. `*Nightly benchmarking CI status*`

If a section is empty, write `(none)`.

## Content

PR entries: `<URL|short-title>` by author: brief summary (5-10 words).

> Closed PRs: "N PRs closed: reason": no URLs.

Author: ```suggestion
Closed PRs: `<URL|short-title>` by author: brief summary (5-10 words).
```

Use the provided Markdown context file (`weekly_slack_summary_context.md`) as the
source of facts.

## Ground rules

- Return the final mrkdwn summary directly in your response.
- Do not create files and do not wrap the output in a code block.
- Use only facts present in the context.
- Do not use em-dashes (`—`); use hyphens (`-`) or colons instead.
- Do not use Markdown headings (`#`, `##`, etc.). Use `*Section Title*` for
  section titles (Slack renders single asterisks as bold).
- Format every link as Slack mrkdwn: `<URL|display text>`. Do not use raw URLs
  or Markdown `[text](url)`.

## Required content for every PR and issue

For each listed item, include all of:

1. Title as a clickable `<URL|title>` link
2. Author (the user who opened it)
3. A one-sentence description of what the item is actually about

The description must not simply restate the title. Use the body, commit
messages, review comments, or issue comments to produce a meaningful summary.
If the title is vague (e.g. "fix bug", "update foo"), explain the concrete
change. If there is no additional information beyond the title, say so.

## Output format

1. `*icon4py weekly summary: <start-date> - <end-date>*`

> 2. `*Closed PRs*`

Reviewer: ```suggestion
2. `*Merged PRs*`
```
because maybe we don't want to mix the merged with the closed but not merged, and probably we don't care in the summary about closed-not-merged PRs.

Author: This is rare, but last week there was:

> - [Test --tach in CI](https://github.com/C2SM/icon4py/pull/1291) by msimberg: Attempted to integrate tach testing into CI, but closed as the monorepo structure complicates the implementation.

I think we can ask the summarizer to separate the categories though.

   - `<PR URL|title>` by <author>: <one-sentence summary>
3. `*Ongoing PRs*`
   - `<PR URL|title>` by <author>: \<one-sentence status/summary>
4. `*Inactive open PRs*`
   - List up to 9 highlighted inactive PRs: 3 newest, 3 oldest, 3 most active.
   - Format: `<PR URL|title>` by <author> \[<reason tags>\]
5. `*Opened issues*`
   - `<issue URL|title>` by <author>: <one-sentence summary>
6. `*Closed issues*`
   - `<issue URL|title>` by <author>: <one-sentence summary>
7. `*Weekly CI status*`
   - Report the status for the current week (Monday to Sunday), even though the
     PRs and issues above are from the previous week.
   - Always include the pipeline URL as a clickable `<URL|pipeline>` link.
   - For failed jobs, group them by failure category (e.g., runner failures,
     test failures, infrastructure failures). Show up to 5 categories. For each
     category, give the total count, one representative clickable
     `<job URL|job name>` link, and a one-sentence summary based on the
     truncated raw log tail. Distinguish CI/runner failures from actual test
     failures.

>    - If there are more than 5 categories, state the total number of categories

Reviewer 2: I rely on the fact that this will also detect performance regression in Bencher nightly CI jobs. I think it should work, but I will keep an eye (performance regressions are very rare, though).

Author: Currently it won't look at bencher pipelines, it'll only look at the `all` pipeline that runs early on monday morning. I'll ask it to check the bencher runs as well (but only the latest run, not the full last week). Do you think that works?

Reviewer 2: Yes, it is enough to only look at the latest run. Thank you
